usr/src/uts/common/fs/zfs/sys/spa.h

```diff
@@ -326,7 +327,8 @@ typedef enum bp_embedded_type {
 typedef struct blkptr {
 	dva_t		blk_dva[SPA_DVAS_PER_BP]; /* Data Virtual Addresses */
 	uint64_t	blk_prop;	/* size, compression, type, etc	    */
-	uint64_t	blk_pad[2];	/* Extra space for the future	    */
+	uint32_t	blk_iv[3];	/* IV for block level encryption    */
+	uint32_t	blk_pad;	/* Extra space for the future	    */
```

Comment:
> Yes, ignoring dedup, seems like the IV could be either the physical location (BP_IDENTITY()) or logical location (the objset, object, and blkid (I'm assuming all encrypted blocks are level=0)).

They are all level 0 (or -2 for ZIL), but the bookmark actually won't work. A snapshot will have a different objset. 

> For dedup, what do you propose to use for the IV? I think a random number wouldn't work because then the encrypted data (and thus checksum) would be different and thus you wouldn't dedup.

Yeah this has the same problem as `BP_IDENTITY()`. I honestly don't have a good idea here (and obviously neither did Oracle when they implemented it in Solaris).

> It seems like if you want dedup, you will have to accept some information leakage. At a minimum, someone reading the raw disks would know where the redundant data is. I can write arbitrary data to files, and read the raw disks, then I can guess and check what your data is, by seeing if it dedup's to the same block as my data.

This is technically true, but I don't think this is a viable threat. The attacker would first need the key to already be loaded into the pool, otherwise the dataset can't write to do dedup in the first place anyway. They would also need write access to the dataset, at which point they can just read the data anyway.